### PR TITLE
overestimation free quadratic expansion

### LINKED ISSUE #109: Optimum term in quadratic expansion
Doing `(α + βx) x`  looses the dependency between the arguments, so it may be less precise using interval arithmetic. If there is a general rule, one has to find it by taking the derivative of the expression as in the paper, which is a particular case with alpha, beta being something like t and 1/2t^2.

```julia
julia> fprod(α, β, x) = (α + β*x)*x
fprod (generic function with 1 method)

julia> fsum(α, β, x) = α*x + β*x^2
fsum (generic function with 1 method)

julia> y = Interval(-1, 1)
[-1, 1]

julia> fprod(0.0, 1.0, y)
[-1, 1]

julia> fsum(0.0, 1.0, y) # fsum is tighter
[0, 1]

julia> y = Interval(-3, -2)
[-3, -2]

julia> fprod(1.0, 2.0, y) # fprod is tighter
[6, 15]

julia> fsum(1.0, 2.0, y)
[5, 16]
```

_Originally posted by @mforets in https://github.com/JuliaReach/IntervalMatrices.jl/pull/91#issuecomment-579806859_

### PR DESCRIPTION
**Pull request by lucaferranti**
fixes #109 

With a quick look at the `exponential.jl` file, that function isn't used anywhere else anyway?

### CODE REVIEW

File: src/exponential.jl [hunk 1]
```diff
@@ -121,6 +121,12 @@ function quadratic_expansion(A::IntervalMatrix, α::Real, β::Real)
     return B
 end
 
+function quadratic_expansion(x::Interval, α::Real, β::Real)
```
Comment (mforets) [inline src/exponential.jl:124]: i suggest to add a basic docstring:
 
````
"""
    quadratic_expansion(x::Interval, α::Real, β::Real)

Compute `(α + βx) x` with a single-use expression.

### Input

...

### Output

...
"""
function quadratic_expansion(x::Interval, α::Real, β::Real)
````
```suggestion
function quadratic_expansion(x::Interval, α::Real, β::Real)
```

Comment (mforets) [reply src/exponential.jl:124]: oh, i take back my comment, given that the docstring for the function just above this one is OK.

**Review by schillic — APPROVED** [review]

**Comment by mforets** [comment]
> With a quick look at the exponential.jl file, that function isn't used anywhere else anyway?

indeed, i didn't spot other places where it is used.

otoh, see for example [here](https://github.com/JuliaReach/ReachabilityAnalysis.jl/blob/master/src/Discretization/exponentiation.jl#L451), method that wants to compute the term `1/2 * δ^2 * A + 1/6 * δ^3 * A²`. so indeed instead of 

```julia
IδW = Iδ + 1/2 * δ^2 * A + 1/6 * δ^3 * A²
```
that method could be improved by using
```julia
IδW = Iδ + quadratic_expansion(A, 1/2 * δ^2, 1/6 * δ^3)
```

File: src/exponential.jl (hunk 1 above)
Comment (lucaferranti) [reply src/exponential.jl:124]: does "single-use" here mean "rewritten so that it doesn't suffer from dependency problem"?

**Review by mforets — APPROVED** [review]

**Comment by mforets** [comment]
@lucaferranti  how about that we add the tests from https://github.com/JuliaReach/IntervalMatrices.jl/issues/109#issuecomment-924191900 ?

File: src/exponential.jl (hunk 1 above)
Comment (schillic) [reply src/exponential.jl:124]: Yes.

Comment (mforets) [reply src/exponential.jl:124]: yes. i read it abbreviated SUE somewhere

**Comment by lucaferranti** [comment]
what about the 

```julia
quadratic_expansion(A::IntervalMatrix{T}, t) where {T}
```

from what I can tell, it's the same of the more general one with $\alpha=t$ $\beta=1/2t^2$, is the current implementation of the 2-input methods overestimate free? maybe that could be replaced with 

```julia
quadratic_expansion(A::IntervalMatrix{T}, t) where {T} = quadratic_expansion(A, t, 1/2*t^2)
```

**Comment by lucaferranti** [comment]
```julia
julia> t = 1.0
1.0

julia> α = 1.0
1.0

julia> β = 0.5
0.5

julia> A = IntervalMatrix(randn(100, 100) .± abs.(randn(100, 100)));

julia> @btime quadratic_expansion($A, $t);
  135.591 ms (7544198 allocations: 115.27 MiB)

julia> @btime quadratic_expansion($A, $α, $β);
  71.674 ms (3606 allocations: 296.91 KiB)
```

so maybe we could define

```julia
quadratic_expansion(A, t) = quadratic_expansion(A, t, t^2/2)
```

**Comment by mforets** [comment]
yes, the code refactoring that you propose sounds good to me

**Comment by lucaferranti** [comment]
quick question about the docstring, it currently says

```
### Input
- `A` -- interval matrix
- `t` -- non-negative time value
```

would it be better to keep it "application agnostic" and say that `t` is just a real number? Mathematically, I think the algorithm works also if `t` is negative (at least after the proposed refactoring)

**Comment by schillic** [comment]
True. Maybe it is better to just write `quadratic_expansion(A, t, t^2/2)` in the two places where it is used and remove that method altogether.

**Comment by lucaferranti** [comment]
the function is exported though, is it used in some other packages? According to JuliaHub, it has zero dependents

**Comment by mforets** [comment]
>  According to JuliaHub, it has zero dependents

hmm  i saw that too ([here](https://juliahub.com/ui/Packages/IntervalMatrices/VerWp/0.7.0?t=2)). strange, since IM is a dependency of ReachabilityAnalysis (see [here](https://juliahub.com/ui/Packages/ReachabilityAnalysis/fYI0z/0.15.1?t=1))

**Comment by mforets** [comment]
about removing `quadratic_expansion(A, t)`, sure, why not. i think this method is only used internally.

the reason to have a special method is that it's the canonical case (for the matrix exp expansion).

**Comment by lucaferranti** [comment]
just double checking one last time before pulling the trigger, remove or replace `quadratic_expansion(A, t)`?

**Review by schillic — APPROVED** [review]

**Comment by lucaferranti** [comment]
added the test for quadratic expansion from the issue comment, now this should be ready.

I noticed the tests for exponentials aren't really testing the result is correct, just that the output is an interval matrix, but I guess that can be addressed in a different PR

**Comment by schillic** [comment]
Yes, we were eager to see results back then :sweat_smile:

**Comment by lucaferranti** [comment]
for the release, since this is breaking and there are at least a couple of other breaking changes coming in the upcoming days (moving exponential to ILA, bringing rump multiplication here, etc.), maybe we can wait till Tuesday meeting for the next release?

**Review by mforets — APPROVED** [review]

**Pull request merged**